content/2015.09.kinto-encryption-example.rst

```diff
@@ -0,0 +1,207 @@
+How to encrypt data with Kinto.js
+###########################################
+
+:date: 2015-09-16
+:summary: How to encrypt data with Kinto.js.
+:slug: kinto-encryption-example
+:lang: en
+:url: kinto-encryption-example
+
+A guest post this time, by Michiel de Jong from the Firefox OS team! Sorry for
+writing in English. :)
+
+If you want to encrypt data that is uploaded to the cloud by an offline-first app,
+it makes sense to keep the local copy of the data in-the-clear, and encrypt the
+data only just before it is uploaded. That way, you can easily and efficiently display
+and update the local copy of the data, without the repeated cost of encrypting
+and decrypting everything all the time. Likewise, you only need to decrypt incoming data
+once, when it's downloaded, and your app can then display the data as often as
+needed without extra processing costs.
+
+Kinto.js now supports
+`<code>RemoteTransformer</code>s http://kintojs.readthedocs.org/en/latest/api/#transformers`_,
+which are superhandy for this pattern.
+This blogpost shows how to use a <code>RemoteTransformer</code> for encrypting your data just
+before it's sent to the cloud. It uses WebCrypto for the cryptography.
+
+The main idea behind a <code>RemoteTransformer</code> is that you provide two functions,
+"encode" and "decode", and Kinto.js makes sure to call the encode function on
+each record that gets uploaded, and to call the decode function on each record
+that gets downloaded. You can also read about this in the Kinto.js documentation,
+http://kintojs.readthedocs.org/.
+
+NB: I am not a security expert and I wrote this example in an afternoon, so
+if you're going to copy-paste this code, please don't use it for anything
+valuable without doing your own proper security review first! :)
+
+The way we tell Kinto.js that we want to use encryption, and provide our own
+"encode" and "decode" procedures, is quite simple. When creating a Kinto
+collection, we pass a <code>RemoteTransformer</code> as a collection-level option. The full
+code example is
+`on github http://michielbdejong.github.io/kinto-encryption-example/app.js`_,
+but let's walk through the code step-by-step.
+
+If you're unfamiliar with ES6, you may be surprised by some of this syntax; so let's
+take this opportunity to get used to it! Here's a little translation:
+* <code>() => {}</code> is similar to <code>function() {}</code>
+(although the scoping of <code>this</code> is different).
+* <code>const foo = 'bar'</code> and <code>let foo = 'bar'</code> are similar to
+<code>var foo = 'bar'</code>.
+* <code>{ a, b }</code> is similar to <code>{ a: a, b: b }</code>.
+* <code>`Hello ${foo}`</code> (note the backticks) is similar to
+<code>'Hello ' + foo</code>.
```

Comment:
Adding a note linking to some ES6 summary of new features might be useful here; eg. http://babeljs.io/docs/learn-es2015/